### PR TITLE
Enable INT8 pool OP

### PR DESCRIPTION
**Pull request by xiaolil1**
Split from #14929
test=develop

### CODE REVIEW

**Comment by luotao1** [comment]
```
[01:35:00][Step 1/1] test_pool2d_int8_mkldnn_op failed
[01:35:00][Step 1/1]  EEEEE
[01:35:00][Step 1/1] ======================================================================
[01:35:00][Step 1/1] ERROR: test_check_output (test_pool2d_int8_mkldnn_op.TestCase1)
[01:35:00][Step 1/1] ----------------------------------------------------------------------
[01:35:00][Step 1/1] Traceback (most recent call last):
[01:35:00][Step 1/1]   File "/paddle/build/python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py", line 151, in test_check_output
[01:35:00][Step 1/1]     self.check_output()
[01:35:00][Step 1/1]   File "/paddle/build/python/paddle/fluid/tests/unittests/op_test.py", line 385, in check_output
[01:35:00][Step 1/1]     self.check_output_with_place(place, atol, no_check_set, equal_nan)
[01:35:00][Step 1/1]   File "/paddle/build/python/paddle/fluid/tests/unittests/op_test.py", line 308, in check_output_with_place
[01:35:00][Step 1/1]     outs, fetch_list = self._calc_output(place, no_check_set=no_check_set)
[01:35:00][Step 1/1]   File "/paddle/build/python/paddle/fluid/tests/unittests/op_test.py", line 300, in _calc_output
[01:35:00][Step 1/1]     return_numpy=False)
[01:35:00][Step 1/1]   File "/paddle/build/python/paddle/fluid/executor.py", line 472, in run
[01:35:00][Step 1/1]     self.executor.run(program.desc, scope, 0, True, True)
[01:35:00][Step 1/1] EnforceNotMet: op pool2d does not have kernel for data_type[int8_t]:data_layout[ANY_LAYOUT]:place[CUDAPlace(0)]:library_type[PLAIN] at [/paddle/paddle/fluid/framework/operator.cc:916]
[01:35:00][Step 1/1] PaddlePaddle Call Stacks: 
[01:35:00][Step 1/1] 0       0x7f107518e07dp void paddle::platform::EnforceNotMet::Init<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char const*, int) + 509
[01:35:00][Step 1/1] 1       0x7f1076d06d85p paddle::framework::OperatorWithKernel::RunImpl(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) const + 1365
[01:35:00][Step 1/1] 2       0x7f1076d00f60p paddle::framework::OperatorBase::Run(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) + 544
[01:35:00][Step 1/1] 3       0x7f1075289bfep paddle::framework::Executor::RunPreparedContext(paddle::framework::ExecutorPrepareContext*, paddle::framework::Scope*, bool, bool, bool) + 302
[01:35:00][Step 1/1] 4       0x7f107528c37ap paddle::framework::Executor::Run(paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool) + 218
[01:35:00][Step 1/1] 5       0x7f107516391ap
[01:35:00][Step 1/1] 6       0x7f10751b687dp
[01:35:00][Step 1/1] 7             0x4c5326p PyEval_EvalFrameEx + 37958
[01:35:00][Step 1/1] 8             0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 9             0x4c17c6p PyEval_EvalFrameEx + 22758
[01:35:00][Step 1/1] 10            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 11            0x4c17c6p PyEval_EvalFrameEx + 22758
[01:35:00][Step 1/1] 12            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 13            0x4c17c6p PyEval_EvalFrameEx + 22758
[01:35:00][Step 1/1] 14            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 15            0x4c17c6p PyEval_EvalFrameEx + 22758
[01:35:00][Step 1/1] 16            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 17            0x4c1f56p PyEval_EvalFrameEx + 24694
[01:35:00][Step 1/1] 18            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 19            0x4d57a3p
[01:35:00][Step 1/1] 20            0x4a587ep PyObject_Call + 62
[01:35:00][Step 1/1] 21            0x4be51ep PyEval_EvalFrameEx + 9790
[01:35:00][Step 1/1] 22            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 23            0x4d5669p
[01:35:00][Step 1/1] 24            0x4eef5ep
[01:35:00][Step 1/1] 25            0x4a587ep PyObject_Call + 62
[01:35:00][Step 1/1] 26            0x548fc3p
[01:35:00][Step 1/1] 27            0x4c166dp PyEval_EvalFrameEx + 22413
[01:35:00][Step 1/1] 28            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 29            0x4d57a3p
[01:35:00][Step 1/1] 30            0x4a587ep PyObject_Call + 62
[01:35:00][Step 1/1] 31            0x4be51ep PyEval_EvalFrameEx + 9790
[01:35:00][Step 1/1] 32            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 33            0x4d5669p
[01:35:00][Step 1/1] 34            0x4eef5ep
[01:35:00][Step 1/1] 35            0x4a587ep PyObject_Call + 62
[01:35:00][Step 1/1] 36            0x548fc3p
[01:35:00][Step 1/1] 37            0x4c166dp PyEval_EvalFrameEx + 22413
[01:35:00][Step 1/1] 38            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 39            0x4d57a3p
[01:35:00][Step 1/1] 40            0x4a587ep PyObject_Call + 62
[01:35:00][Step 1/1] 41            0x4be51ep PyEval_EvalFrameEx + 9790
[01:35:00][Step 1/1] 42            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 43            0x4d5669p
[01:35:00][Step 1/1] 44            0x4eef5ep
[01:35:00][Step 1/1] 45            0x4a587ep PyObject_Call + 62
[01:35:00][Step 1/1] 46            0x548fc3p
[01:35:00][Step 1/1] 47            0x4c166dp PyEval_EvalFrameEx + 22413
[01:35:00][Step 1/1] 48            0x4c141fp PyEval_EvalFrameEx + 21823
[01:35:00][Step 1/1] 49            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 50            0x4c1f56p PyEval_EvalFrameEx + 24694
[01:35:00][Step 1/1] 51            0x4b9b66p PyEval_EvalCodeEx + 774
[01:35:00][Step 1/1] 52            0x4eb69fp
[01:35:00][Step 1/1] 53            0x4e58f2p PyRun_FileExFlags + 130
[01:35:00][Step 1/1] 54            0x4e41a6p PyRun_SimpleFileExFlags + 390
[01:35:00][Step 1/1] 55            0x4938cep Py_Main + 1358
[01:35:00][Step 1/1] 56      0x7f10c73a9830p __libc_start_main + 240
[01:35:00][Step 1/1] 57            0x493299p _start + 41
[01:35:00][Step 1/1] 
[01:35:00][Step 1/1] 
[01:35:00][Step 1/1] ======================================================================
```
http://ci.paddlepaddle.org/viewLog.html?buildId=41797&tab=buildLog&buildTypeId=Paddle_PrCi&logTab=tree&filter=all#_state=62

File: python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+from __future__ import division
+
+import unittest
+import numpy as np
+
+import paddle.fluid.core as core
+from op_test import OpTest
+
+
+def adaptive_start_index(index, input_size, output_size):
```
Comment (luotao1) [inline python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py]: These four function `adaptive_start_index, adaptive_end_index, max_pool2D_forward_naive, avg_pool2D_forward_naive` are the same in `test_pool2d_op.py`. 
**Please do not copy it.** And it's not need in this PR.

File: python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                c_end = np.min((j * strides[1] + ksize[1] - paddings[1], W))
+            x_masked = x[:, :, r_start:r_end, c_start:c_end]
+
+            field_size = ((r_end - r_start) * (c_end - c_start)) \
+                        if (exclusive or adaptive) else (ksize[0] * ksize[1])
+            out[:, :, i, j] = np.sum(x_masked, axis=(2, 3)) / field_size
+    return out
+
+
+class TestPool2D_Op(OpTest):
```
Comment (luotao1) [inline python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py]: `TestPool2D_Op` is not needed. You can inherit from `test_pool2d_py`. See examples in `test_batch_norm_mkldnn.py`.

https://github.com/PaddlePaddle/Paddle/blob/9c2cbfb89e8cb30aa8d8e12ac7af42fec9f7a9d3/python/paddle/fluid/tests/unittests/test_batch_norm_mkldnn_op.py#L27

File: python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        self.strides = [1, 1]
+        self.paddings = [0, 0]
+        self.dtype = np.int8
+
+    def init_pool_type(self):
+        self.pool_type = "avg"
+        self.pool2D_forward_naive = avg_pool2D_forward_naive
+
+    def init_global_pool(self):
+        self.global_pool = False
```
Comment (luotao1) [inline python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py:53]: line178-190 are not needed. For `self.dtype = np.int8`, you can see following examples:
```
def init_kernel_type(self):
   self.dtype = np.int8
```
https://github.com/PaddlePaddle/Paddle/blob/9c2cbfb89e8cb30aa8d8e12ac7af42fec9f7a9d3/python/paddle/fluid/tests/unittests/test_batch_norm_mkldnn_op.py#L28-L29
And update `test_pool2d_py` like
https://github.com/PaddlePaddle/Paddle/blob/9c2cbfb89e8cb30aa8d8e12ac7af42fec9f7a9d3/python/paddle/fluid/tests/unittests/test_batch_norm_op.py#L289

File: python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    def init_test_case(self):
+        self.shape = [2, 3, 7, 7]
+        self.ksize = [3, 3]
+        self.strides = [1, 1]
+        self.paddings = [1, 1]
+        self.dtype = np.uint8
+
+    def init_pool_type(self):
+        self.pool_type = "max"
+        self.pool2D_forward_naive = max_pool2D_forward_naive
```
Comment (luotao1) [inline python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py:82]: There are 6 pool tests in `test_pool2d_op`. **For `int8` and `uint8`, there should be 12 pool tests here.** You can follow `test pool2d cudnn` examples to make the code more clean and readable. **Please update `test_pool2d_mkldnn_op.py` at the same time.**
https://github.com/PaddlePaddle/Paddle/blob/9c2cbfb89e8cb30aa8d8e12ac7af42fec9f7a9d3/python/paddle/fluid/tests/unittests/test_pool2d_op.py#L248-L265

**Review by luotao1 — COMMENTED** [review]
Since there is only one `pool_mkldnn_op`, please combine `test_pool2d_int8_mkldnn_op.py` into `test_pool2d_mkldnn_op.py`.

File: python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py (hunk 1 above)
Comment (xiaolil1) [reply python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py]: done.

File: python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py (hunk 2 above)
Comment (xiaolil1) [reply python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py]: done.

File: python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py (hunk 3 above)
Comment (xiaolil1) [reply python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py:53]: done.

File: python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py (hunk 4 above)
Comment (xiaolil1) [reply python/paddle/fluid/tests/unittests/test_pool2d_int8_mkldnn_op.py:82]: done.

**Comment by xiaolil1** [comment]
Failed to perform checkout on agent: '"C:\Program Files\Git\bin\git.exe" clean -f -d -x' command failed.
exit code: 1
stderr: warning: Could not stat path 'build/paddle/fluid/framework/details/Release/broadcast_op_test.exe': Permission denied
warning: Could not stat path 'build/paddle/fluid/framework/ir/Release/graph_helper_test.exe': Permission denied
warning: Could not stat path 'build/paddle/fluid/framework/Release/cow_ptr_tests.exe': Permission denied
warning: Could not stat path 'build/paddle/fluid/operators/jit/Release/jit_kernel_test.exe': Permission denied
warning: Could not stat path 'build/paddle/fluid/operators/math/Release/sequence_pooling_test.exe': Permission denied
warning: Could not stat path 'build/paddle/fluid/platform/Release/temporal_allocator_test.exe': Permission denied
warning: Could not stat path 'build/third_party/install/mklml/lib/mklml.dll': Permission denied

**Review by tensor-tang — APPROVED** [review]
LGTM

**Pull request merged**